scripts/alert_emails/create_lists_to_email.ts

```diff
@@ -1,69 +1,95 @@
- *     - snapshots/<id>/locations/<fips>/emails/<email> => {sent: null }
+ * Takes the alerts file generated by `generate_daily_alerts.ts` and updates the
+ * list of emails that we need to send for each location. The steps are:
+ *
+ * 1. Read the locations that changed threat level from the alerts file.
+ * 2. Fetch the emails that are subscribed to the locations that changed threat level.
+ * 3. Creates a list of emails for each location and stores them in Firestore.
  *
- * Takes in a file with {[fips: string]: {...data}} and generates documents
- * for each user to be emailed.
+ * Run the script via yarn:
  *
- * You can run via `yarn create-lists-to-email fipsToAlertFilename currentSnapshot`
+ *   $ yarn create-lists-to-email alerts.json snapshotId
  */
-(async () => {
-  const outputFolder = path.join(__dirname);
-  await fs.ensureDir(outputFolder);
+const updateSnapshotEmails = async (
+  alertsFileName: string,
+  snapshotId: string,
+) => {
+  const db = getFirestore();
+  const alertsByLocation = readAlertsFile(alertsFileName);
+  const fipsList = Object.keys(alertsByLocation);
 
-  const { fipsToAlertFilename, currentSnapshot } = await parseArgs();
-  const alertPath = path.join(outputFolder, fipsToAlertFilename);
+  fipsList.forEach(async (fips: string) => {
+    const emailQuerySnapshot = await db
+      .collection('alerts-subscriptions')
+      .where('locations', 'array-contains', fips)
+      .get();
 
-  const rawdata = fs.readFileSync(alertPath, 'utf8');
-  const locationsWithAlerts: { [fips: string]: Alert } = JSON.parse(rawdata);
+    // Remove invalid and repeated emails for the location from the list
+    const emailList = _.uniq(
+      emailQuerySnapshot.docs
+        .map((emailDoc: { id: string }) => emailDoc.id)
+        .filter(isValidEmail)
+        .map(_.toLower),
+    );
 
-  const db = getFirestore();
-  Object.keys(locationsWithAlerts).forEach(async (fips: string) => {
-    db.collection('alerts-subscriptions')
-      .where('locations', 'array-contains', fips)
-      .get()
-      .then(querySnapshot => {
-        querySnapshot.forEach(doc => {
-          db.collection('snapshots')
-            .doc(`${currentSnapshot}/locations/${fips}/emails/${doc.id}`)
-            .set({
-              sentAt: null,
-            });
-        });
-        return;
-      })
-      .catch((err: any) => {
-        console.log(err);
-      });
+    try {
+      await Promise.all(
+        emailList.map(async (email: string) => {
+          try {
+            await db
+              .collection('snapshots')
+              .doc(`${snapshotId}/locations/${fips}/emails/${email}`)
+              .create({ sentAt: null });
+          } catch (updateError) {
+            // We don't want to overwrite existing emails for a location to avoid
+            // double-sending alerts, in case this is a re-run of the send-alert-emails
+            if (updateError.code !== ERROR_CODE_ALREADY_EXISTS) {
```

Comment:
Is it worth doing a `console.error('Firestore write failed', updateError)` as well?  In particular if Promise.all() doesn't surface multiple errors?